rpcs3/Emu/RSX/Overlays/overlay_perf_metrics.cpp

```diff
@@ -0,0 +1,205 @@
+#include "stdafx.h"
+#include "overlays.h"
+#include "../GSRender.h"
+
+#include "Emu/Cell/SPUThread.h"
+#include "Emu/Cell/RawSPUThread.h"
+#include "Emu/Cell/PPUThread.h"
+#include "Utilities/sysinfo.h"
+
+namespace rsx
+{
+	namespace overlays
+	{
+		void perf_metrics_overlay::reset_body()
+		{
+			m_body.set_font("n023055ms.ttf", m_font_size);
+			m_body.set_pos(50, 50);
+			m_body.fore_color     = {0xFF / 255.f, 0xe1 / 255.f, 0x38 / 255.f, 1.0f};
+			m_body.back_color     = {0x00 / 255.f, 0x23 / 255.f, 0x39 / 255.f, 0.7f};
+			m_body.padding_top    = 0;
+			m_body.padding_bottom = -14;
+			m_body.set_margin(5, 5, 0, 0);
+		}
+
+		void perf_metrics_overlay::reset_titles()
+		{
+			m_titles.set_font("n023055ms.ttf", m_font_size);
+			m_titles.set_pos(50, 50);
+			m_titles.fore_color     = {0xf2 / 256.0, 0x6C / 256.0, 0x24 / 256.0, 1.0f};
+			m_titles.back_color     = {0.0f, 0.0f, 0.0f, 0.0f};
+			m_titles.padding_top    = 0;
+			m_titles.padding_bottom = -14;
+			m_titles.set_margin(5, 5, 0, 0);
+
+			switch (m_detail)
+			{
+			case detail_level::minimal:
+			case detail_level::low: m_titles.text = ""; break;
+			case detail_level::medium: m_titles.text = fmt::format("\n\n%s", title1_medium); break;
+			case detail_level::high: m_titles.text = fmt::format("\n\n%s\n\n\n\n\n%s", title1_high, title2); break;
+			}
+			m_titles.auto_resize();
+			m_titles.refresh();
+		}
+
+		void perf_metrics_overlay::init()
+		{
+			reset_titles();
+			reset_body();
+
+			update(true);
+			m_update_timer.Start();
+		}
+
+		perf_metrics_overlay::perf_metrics_overlay(bool initialize)
+		{
+			// Default values, will change based on config options
+			m_update_interval = 350;
+			m_detail          = detail_level::high;
+			m_font_size       = 11;
+
+			if (initialize)
+				init();
+		}
+
+		// In ms
+		void perf_metrics_overlay::set_update_interval(u32 update_interval)
+		{
+			m_update_interval = update_interval;
+		}
+
+		void perf_metrics_overlay::set_detail_level(detail_level level)
+		{
+			m_detail = level;
+			reset_titles();
+		}
+
+		void perf_metrics_overlay::set_font_size(u32 font_size)
+		{
+			m_font_size = font_size;
+			reset_titles();
+			reset_body();
+		}
+
+		void perf_metrics_overlay::update(bool force_update)
+		{
+			const auto elapsed = m_update_timer.GetElapsedTimeInMilliSec();
+
+			if (!force_update)
+			{
+				++m_frames;
+			}
+
+			if (elapsed >= m_update_interval || force_update)
+			{
+				f32 fps{0};
+				f32 frametime{0};
+
+				u64 ppu_cycles{0};
+				u64 spu_cycles{0};
+				u64 rsx_cycles{0};
+				u64 total_cycles{0};
+
+				u32 ppus{0};
+				u32 spus{0};
+				u32 rawspus{0};
+
+				f32 cpu_usage{-1.f};
+				u32 total_threads{0};
+
+				f32 ppu_usage{0};
+				f32 spu_usage{0};
+				f32 rsx_usage{0};
+				u32 rsx_load{0};
+
+				std::shared_ptr<GSRender> rsx_thread;
+
+				std::string perf_text;
+
+				// 1. Fetch/calculate metrics we'll need
+				switch (m_detail)
+				{
```

Comment:
With this switch block and the one after it, it might be easier on the eyes if the cases were braced, because they have a lot of content spanning many lines in some of the blocks.